src/models/state/mod.rs

```diff
@@ -1472,6 +1482,92 @@ impl GlobalState {
         // Ok(())
     }
 
+    /// Returns (parent, child) pair of blocks.
+    pub(crate) async fn fetch_block_pair(&self, child_digest: Digest) -> Option<(Block, Block)> {
+        let child = self
+            .chain
+            .archival_state()
+            .get_block(child_digest)
+            .await
+            .expect("fetching block from archival state should work.");
+        let Some(child) = child else {
+            warn!("Got sync challenge for unknown tip");
+
+            return None;
+        };
+        let parent_digest = child.header().prev_block_digest;
+        let parent = self
+            .chain
+            .archival_state()
+            .get_block(parent_digest)
+            .await
+            .expect("fetching block from archival state should work.")
+            .expect("parent of known block from archival state must exist.");
+
+        Some((parent, child))
+    }
+
+    pub(crate) async fn response_to_sync_challenge(
+        &self,
+        sync_challenge: SyncChallenge,
+    ) -> Result<SyncChallengeResponse> {
+        let Some((tip_parent, tip)) = self.fetch_block_pair(sync_challenge.tip_digest).await else {
+            bail!("could not fetch tip and tip predecessor");
+        };
+
+        if tip.header().height < MIN_BLOCK_HEIGHT_FOR_SYNCING.into() {
+            bail!("tip height is too small for sync mode")
+        }
+
+        let mut block_pairs: Vec<(TransferBlock, TransferBlock)> = vec![];
+        let mut block_mmr_mps = vec![];
+        for h in sync_challenge.challenges {
+            if h < 2u64.into() {
+                bail!("challenge asks for genesis block");
+            }
+
+            let Some(child_digest) = self
+                .chain
+                .archival_state()
+                .archival_block_mmr
+                .try_get_leaf(h.into())
+                .await
+            else {
+                bail!("could not get leaf from archival block mmr");
+            };
+            let Some((p, c)) = self.fetch_block_pair(child_digest).await else {
+                bail!("could not fetch indicated block pair");
+            };
+
+            block_mmr_mps.push(
```

Comment:
If `tip` in request does not match own tip, then these MMR membership proofs will not be valid. Is that OK? We might as well check for that here and then decide what to do in that case.

Reply:
Addressed with comment in function.